extensions/notification/NotificationConfigManager.js

```diff
@@ -0,0 +1,402 @@
+const async = require('async');
+const { EventEmitter } = require('events');
+const zookeeper = require('node-zookeeper-client');
+const joi = require('@hapi/joi');
+
+const safeJsonParse = require('./utils/safeJsonParse');
+const constants = require('./constants');
+
+const paramsJoi = joi.object({
+    zkClient: joi.object().required(),
+    parentNode: joi.string().optional(),
+    logger: joi.object().required(),
+}).required();
+
+/**
+ * @class NotificationConfigManager
+ *
+ * @classdesc Manages bucket notification configurations in zookeeper, maintains
+ * configurations across multiple managers.
+ */
+class NotificationConfigManager {
+    /**
+     * @constructor
+     * @param {Object} params - constructor params
+     * @param {Object} params.zkClient - zookeeper client
+     * @param {String} params.parentNode - parent node to be used
+     * @param {Logger} params.logger - logger object
+     */
+    constructor(params) {
+        joi.attempt(params, paramsJoi);
+        this._zkClient = params.zkClient;
+        this._parentNode = params.parentNode || constants.zkConfigParentNode;
+        this.log = params.logger;
+        this._configs = new Map();
+        this._concurrency = constants.configManager.concurrency;
+        this._emitter = new EventEmitter();
```

Comment:
Yes, I did consider it. Though, while implementing something made me not to. I can't remember what or why. Will change it.

Reply:
The events are emitted only for internal(private) purposes of the class, not intended for the instances of the class to listen to directly. May be its not the cleanest implementation, do you recommend a different approach?

Reply:
My take: if the events are meaningful you don't necessarily need to prevent external users from listening to them. You can sure use them it it helps for the implementation and you can have "internal only" events as well if you need, I don't think it warrants to create a hidden object.

Events are a useful abstraction to decouple emitters and receivers, making the code more modular. Sometimes it's the good approach, sometimes it can be better to just have registered callbacks if the decoupling is not needed, for example if all you need is systematically trigger a single task when something completes. I don't have a special recommendation in this particular implementation.